### PR TITLE
Define some pixel formats

### PR DESCRIPTION
**Pull request by padenot**
Please have a look. Which other formats do we need to support initially (= what formats does Chrome plan to ship with) ? I know there was mentions of NV12.

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/webcodecs/pull/288.html" title="Last updated on Jul 15, 2021, 2:07 PM UTC (81759cc)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webcodecs/288/3584fb9...81759cc.html" title="Last updated on Jul 15, 2021, 2:07 PM UTC (81759cc)">Diff</a>

### CODE REVIEW

**Comment by dalecurtis** [comment]
Here's our current list of formats:
https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/modules/webcodecs/video_pixel_format.idl

ImageDecoder and canvas need I422, I444, and various RGB flavors.

**Comment by sandersdan** [comment]
FYI corrected pixel format names are in [Chromium code review 2946350](https://chromium-review.googlesource.com/c/chromium/src/+/2946350/7/third_party/blink/renderer/modules/webcodecs/video_pixel_format.idl), merge is pending a decision on `hasAlpha` vs format names.

**Comment by chcunningham** [comment]
Definition wise this LGTM. Structurally, do you want to create definition links for some terms here so we can establish references like https://github.com/w3c/webcodecs/pull/208/files/955111c0b90eead2fd1975a33058dcd8e87787ca#r649890397?

**Comment by padenot** [comment]
> FYI corrected pixel format names are in [Chromium code review 2946350](https://chromium-review.googlesource.com/c/chromium/src/+/2946350/7/third_party/blink/renderer/modules/webcodecs/video_pixel_format.idl), merge is pending a decision on `hasAlpha` vs format names.

Thanks for this, makes it very easy.

**Comment by padenot** [comment]
> Definition wise this LGTM. Structurally, do you want to create definition links for some terms here so we can establish references like https://github.com/w3c/webcodecs/pull/208/files/955111c0b90eead2fd1975a33058dcd8e87787ca#r649890397?

Yes, as discussed, I'll see how this is best structured when we have more than one definition.

**Comment by padenot** [comment]
Here are all the pixel formats in @dalecurtis's list in the Chromium patch above, please have a look. I've included a "diagram" for NV12 because the hybrid planar/interleaved makes it slightly unusual maybe, but there is another issue about having proper diagrams.

After we agree on those definitions, I'll add definitions and links as @chcunningham says in https://github.com/w3c/webcodecs/pull/288#issuecomment-868206784.

File: index.src.html [hunk 1]
```diff
@@ -3317,27 +3317,234 @@
...
+    This format is composed of three distinct planes, one plane of Luma and two
+    planes of Chroma, denoted Y, U and V, and present in this order. It is also
+    often refered to as Planar YUV 4:2:0.
+
+    The U an V planes are sub-sampled horizontaly and vertically by a factor of
+    2 compared to the Y planes.
+
+    Each sample in this format is 8 bits.
+
+    There is {{VideoFrame/codedWidth}} * {{VideoFrame/codedHeight}} samples (and
```
Comment (chcunningham) [inline index.src.html:3357]: nit: s/is/are/, here and below

File: index.src.html [hunk 2]
```diff
@@ -3317,27 +3317,234 @@
...
+    image, in {{VideoFrame/codedHeight}} lines of {{VideoFrame/codedWidth}}
+    samples.
+
+    There is {{VideoFrame/codedWidth}} * {{VideoFrame/codedHeight}} / 4 samples
+    (and therefore bytes) in the two U and V planes, arranged starting at the
+    top left in the image, in {{VideoFrame/codedHeight}} / 2 lines of
+    {{VideoFrame/codedWidth}} / 2 samples.
+
+    The {{VideoFrame/codedWidth}} and {{VideoFrame/codedHeight}} MUST be even.
+    Similarly, the crop coordinates ({{VideoFrame/displayWidth}} and
```
Comment (chcunningham) [inline index.src.html:3368]: s/crop coordinates/visible rectangle/ (we used to call them crop)

File: index.src.html [hunk 3]
```diff
@@ -3317,27 +3317,234 @@
...
+    {{VideoFrame/displayHeight}}) MUST be even.
   </dd>
-</dl>
+  <dt><dfn enum-value for=PixelFormat>I420A</dfn></dt>
+  <dd>
+
+    This format is composed of four distinct planes, one plane of Luma, two
+    planes of Chroma, denoted Y, U and V, and one place of alpha values, all
+    present in this order. It is also often refered to as Planar YUV 4:2:0.
```
Comment (chcunningham) [inline index.src.html:3376]: "also often referred to" matches what was written for I420. Should it mention alpha somewhere?

**Review by chcunningham — COMMENTED** [review]
Just nits. Defs and diagram look good.

**Comment by padenot** [comment]
Here are updated commits.

**Review by chcunningham — APPROVED** [review]
LGTM

> After we agree on those definitions, I'll add definitions and links as @chcunningham says in #288 (comment).

Added a few comments to explore this. Can be done here or in a follow up PR if you prefer.

**Comment by chcunningham** [comment]
Aside: I notice you force-pushed. I'm not a git pro ... curious: does that offer some advantage?

**Comment by padenot** [comment]
> Aside: I notice you force-pushed. I'm not a git pro ... curious: does that offer some advantage?

Keeping the patch series as clean as possible for merging, essentially, so then when we'll do git log archeology to understand why we did something, it's not really painful.

File: index.src.html [hunk 4]
```diff
@@ -3317,27 +3317,235 @@
...
   <dd>
-    Planar 4:2:0 YUV.
+    This format is composed of three distinct planes, one plane of Luma and two
+    planes of Chroma, denoted Y, U and V, and present in this order. It is also
+    often refered to as Planar YUV 4:2:0.
+
+    The U an V planes are sub-sampled horizontaly and vertically by a factor of
+    2 compared to the Y planes.
+
+    Each sample in this format is 8 bits.
```
[the remark this answers is not in the document]
Comment (padenot) [reply index.src.html:3355]: Sometimes they are 10 or 12 bits, with 6 or 4 bytes set to zero (even though for now the spec doesn't have this). Maybe we can say that those occupy one byte and all the bits of this byte are meaningful?

File: index.src.html [hunk 5]
```diff
@@ -3317,27 +3317,235 @@
...
 
 <dl>
   <dt><dfn enum-value for=PixelFormat>I420</dfn></dt>
   <dd>
-    Planar 4:2:0 YUV.
+    This format is composed of three distinct planes, one plane of Luma and two
+    planes of Chroma, denoted Y, U and V, and present in this order. It is also
+    often refered to as Planar YUV 4:2:0.
+
+    The U an V planes are sub-sampled horizontaly and vertically by a factor of
```
[the remark this answers is not in the document]
Comment (padenot) [reply index.src.html:3352]: I'm adding a paragraph that defines sub-sampling.

**Comment by padenot** [comment]
I've added more info, let me know if the new paragraph (last patch) works for you. I've consolidated the algorithm you linked to so that it uses common concepts.

File: index.src.html [hunk 6]
```diff
@@ -3317,27 +3317,250 @@
...
 <dl>
   <dt><dfn enum-value for=PixelFormat>I420</dfn></dt>
   <dd>
-    Planar 4:2:0 YUV.
+    This format is composed of three distinct planes, one plane of Luma and two
+    planes of Chroma, denoted Y, U and V, and present in this order. It is also
+    often refered to as Planar YUV 4:2:0.
+
+    The U an V planes are [=sub-sampled=] horizontaly and vertically by a
+    [=factor=] of 2 compared to the Y planes.
```
Comment (chcunningham) [inline index.src.html:3368]: nit: Y plane (no s)

File: index.src.html [hunk 7]
```diff
@@ -3317,27 +3317,250 @@
...
+    {{VideoFrame/displayHeight}}) MUST be even.
+  </dd>
+  <dt><dfn enum-value for=PixelFormat>I422</dfn></dt>
+  <dd>
+
+    This format is composed of three distinct planes, one plane of Luma and two
+    planes of Chroma, denoted Y, U and V, and present in this order. It is also
+    often refered to as Planar YUV 4:2:2.
+
+    The U an V planes are [=sub-sampled=] horizontaly by a [=factor=] of 2
```
Comment (chcunningham) [inline index.src.html:3420]: Here and below, I would make explicit mention of the cases where the factor is 1.

**Review by chcunningham — APPROVED** [review]
Looks great. Just nits. 

> Keeping the patch series as clean as possible for merging, essentially, so then when we'll do git log archeology to understand why we did something, it's not really painful.

I follow. One option: we could change the default merge option on this repo to be "squash" so PRs come it as just one commit. Also fine to just leave it and do force pushes if you like that (offers more flexibility).

**Pull request merged**